Try to make clearTags() more atomic

aaukt opened the pull request:
| Question | Answer |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | php-cache/issues#74 |
| License | MIT |
| Doc PR |  |
### Description

This is my proposed fix for php-cache/issues#74. I'm not trying to avoid any exception as i am not sure that it would help, instead i'm trying to make the clearTags() method a bit more atomic.

To avoid  tagged items in the cache that are not associated to a tag-list anymore, the deletion of the item should be as close to the deletion from the list as possible and not in to seperate steps to first remove all keys from the lists and afterwards remove all keys.
Also as the clearTags() operation is not atomic the cleared list should not be removed as in the meantime new keys might have been added to the list.

aaukt committed:
Try to make clearTags() more atomic

To avoid  tagged items in the cache that are not associated to a tag-list anymore, the deletion of the item should be as close to the deletion from the list as possible and not in to seperate steps to first remove all keys from the lists and afterwards remove all keys.
Also as the clearTags() operation is not atomic the cleared list should not be removed as in the meantime new keys might have been added to the list.

aequasi commented:
I do like this. I am curious though, would it be better to use bulk deletes for the mechanisms that support it (like Redis)?

aaukt commented:
@aequasi But then you have the same situation again, that you are first deleting items from the list and after that delete the items in bulk. If the former fails/is interrupted you have orphaned items in the cache.

Nyholm commented:
This is an important issue. It was highlighted from Andre Romcke as well. Could I bother you to rebase it on master and add some tests to this? 

Note that tagging is in src/Adapter/Common now

aaukt commented:
I rebased and updated the changes.
The modifications to the test would have to be made in cache/integration-tests, how would i go about this to have it available in the branch?

Also i think the tests (testInvalidateTag, testInvalidateTags) currently do not check whether the list is deleted or not even if it says so (it should check for the existence of an item with key "__tag.$TAG").

Nyholm commented on src/Taggable/TaggablePSR6PoolAdapter.php:
```diff
@@ -261,13 +264,6 @@ public function invalidateTags(array $tags)
-        if ($success) {
-            // Remove the tag list
-            foreach ($tags as $tag) {
-                $this->removeList($this->getTagKey($tag));
```
We cannot remove this. The TaggablePSR6PoolAdapter must make sure to manage all the tags itself. It cannot trust the actual `$this->cachePool`

Nyholm commented on src/Adapter/Common/AbstractCachePool.php:
```diff
@@ -351,14 +351,6 @@ public function invalidateTags(array $tags)
         // Remove all items with the tag
         $success = $this->deleteItems($itemIds);
 
-        if ($success) {
```
Thank you

Nyholm reviewed (commented):
I like your changes on the AbstractCachePool but we cannot do the same on the TaggablePSR6PoolAdapter. 

In both cases I think we have to make sure that:
1) We get the tags
2) We remove the item
3) We remove the tags. 

Right, that will make it move fail proof.

Nyholm commented:
> The modifications to the test would have to be made in cache/integration-tests, how would i go about this to have it available in the branch?

If it is a functional test (that does not require any knowledge of our implementation only the taggable "soon-to-be" PSR. Yes make a PR to `cache/integration-tests`. You could do it now and I'll update the tests and create a new release so we can use the new tests in this PR. 

>Also i think the tests (testInvalidateTag, testInvalidateTags) currently do not check whether the list is deleted or not even if it says so (it should check for the existence of an item with key "__tag.$TAG").

Maybe you are correct. That sounds like a unit test, right? If so, add it to src/Adapter/Common/Tests
